File: roc_app/GL/GLVertexArray.cpp
#include "stdafx.h"

#include "GL/GLVertexArray.h"

GLuint GLVertexArray::ms_activeName = 0U;

GLVertexArray::GLVertexArray()
{
    m_name = 0U;
}

bool GLVertexArray::Create()
{
    if(m_name == 0U) glGenVertexArrays(1, &m_name);

    return (m_name != 0U);
}

bool GLVertexArray::Destroy()
{
    if(m_name != 0U)
    {
        if(ms_activeName == m_name) ms_activeName = 0U;
        glDeleteVertexArrays(1, &m_name);
        m_name = 0U;
    }
    return (m_name == 0U);
}

bool GLVertexArray::EnableAttribute(GLuint p_index, GLsizei p_size, GLenum p_type)
{
    if(m_name != 0U)
    {
        if(ms_activeName != m_name) glBindVertexArray(m_name);

        glEnableVertexAttribArray(p_index);
        switch(p_type)
        {
            case GL_FLOAT:
                glVertexAttribPointer(p_index, p_size, p_type, GL_FALSE, 0, NULL);
                break;
            case GL_INT:
                glVertexAttribIPointer(p_index, p_size, p_type, 0, NULL);
                break;
        }

        if(ms_activeName != m_name) glBindVertexArray(ms_activeName);
    }

    return (m_name != 0U);
}

bool GLVertexArray::Bind()
{
    if((m_name != 0) && (ms_activeName != m_name))
    {
        glBindVertexArray(m_name);
        ms_activeName = m_name;
    }

    return (m_name != 0U);
}

bool GLVertexArray::Draw(GLenum p_mode, GLsizei p_count)
{
    if(m_name != 0U)
    {
        if(ms_activeName != m_name) glBindVertexArray(m_name);
        glDrawArrays(p_mode, 0, p_count);
        if(ms_activeName != m_name) glBindVertexArray(ms_activeName);
    }
    return (m_name != 0U);
}

File: roc_app/Engine/OGL/GLVertexArray.cs
using OpenGL;
using System;

namespace ROC.Engine.OGL
{
    internal sealed class GLVertexArray
    {
        [ThreadStatic]
        static uint ms_activeID = 0U;

        uint m_internalID = 0U;

        public GLVertexArray()
        {
            m_internalID = 0U;
        }

        public bool IsValid => (m_internalID != 0U);

        public void Create()
        {
            if(m_internalID != 0U)
                return;

            m_internalID = Gl.GenVertexArray();
        }

        public void Destroy()
        {
            if(m_internalID == 0U)
                return;

            if(ms_activeID == m_internalID)
                ms_activeID = 0U;

            Gl.DeleteVertexArrays(m_internalID);
            m_internalID = 0U;
        }

        public void EnableAttribute(uint p_index, int p_size, int p_type)
        {
            if(m_internalID == 0U)
                return;

            Bind();
            Gl.EnableVertexAttribArray(p_index);

            switch(p_type)
            {
                case Gl.FLOAT:
                    Gl.VertexAttribPointer(p_index, p_size, VertexAttribType.Float, false, 0, IntPtr.Zero);
                    break;

                case Gl.INT:
                    Gl.VertexAttribIPointer(p_index, p_size, VertexAttribType.Int, 0, IntPtr.Zero);
                    break;
            }
        }

        public void Bind()
        {
            if(m_internalID == 0U)
                return;

            if(ms_activeID == m_internalID)
                return;

            Gl.BindVertexArray(m_internalID);
            ms_activeID = m_internalID;
        }

        public void Draw(PrimitiveType p_mode, int p_count)
        {
            if(m_internalID == 0U)
                return;

            Bind();
            Gl.DrawArrays(p_mode, 0, p_count);
        }
    }
}
